canine/adapters/base.py

```diff
@@ -3,6 +3,43 @@
 from itertools import product, repeat
 from functools import reduce
 
+class _FixedArray(object):
+    """
+    Helper to capture arrays which are marked as fixed
+    """
+    def __init__(self, items):
+        self.items = items
+
+    @property
+    def is_2d(self):
+        return len(self.items) > 0 and isinstance(self.items[0], list)
+
+    def __len__(self):
+        return len(self.items) if self.is_2d else 1
+
+    def __iter__(self):
+        if not self.is_2d:
+            raise ValueError("FixedArray is not 2d")
+        for elem in self.items:
+            if isinstance(elem, list):
+                yield _FixedArray(elem)
+            else:
+                yield elem
+
+    def __getitem__(self, n):
+        if self.is_2d and len(self) > n:
+            elem = self.items[n]
+            if isinstance(elem, list):
+                return _FixedArray(elem)
+            return elem
+        raise ValueError("FixedArray is not 2d")
+
+    def stringify(self):
+        return [
+            [str(item) for item in elem] if self.is_2d else str(elem)
+            for elem in self.items
```

Comment:
This will break if the user provided an input containing both lists and scalars.